doc/developer/design/20210505_kafka_keys_in_sql.md

```diff
@@ -0,0 +1,165 @@
+
+## Goals
+
+<!--
+// Enumerate the concrete goals that are in scope for the project.
+-->
+
+Allowing users to specify that a Key should be part of the dataflow, and
+accessible to all SQL processing.
+
+## Non-Goals
+
+<!--
+// Enumerate potential goals that are explicitly out of scope for the project
+// ie. what could we do or what do we want to do in the future - but are not doing now
+-->
+
+* Integrating Key descriptions more tightly into planning or index selection.
+* Any specific optimizations
+
+### Related *future* work
+
+We have two other projects that we know that we would eventually like to
+implement that may have interactions with this design:
+
+* We want to be able to specify a subset of fields from values as the keys for
+  upsert -- although current thinking is that this should probably be done as an
+  upsert function.
+* We would like to expose message metadata as columns or records to the SQL.
+
+This project does not aim to resolve any aspect of those future projects, but
+anything that conflicts with those goals in this design should definitely be
+called out.
+
+## Description
+
+<!--
+// Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+// If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+### Syntax and user-visible semantics
+
+Enhance the Kafka create source syntax like so:
+
+```sql
+CREATE SOURCE <source-name> FROM KAFKA BROKER '<>'
+<format>
+[INCLUDE KEY [AS RECORD <key-record-name>]]
+```
+
+where `<format>` is either:
+
+* An explicit key/value format:
+
+  ```sql
+  KEY FORMAT <specifier>
+  VALUE FORMAT <specifier>
+  ```
+
+* A bare format that specifies a Confluent Schema Registry (CSR) config
+
+  ```sql
+  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY ...
+  ```
+
+If `INCLUDE KEY` is specified, the source schema will be augmented to prepend
+all of the values of the key as columns in the dataflow. Format types that do
+not specify subfields (e.g. `FORMAT TEXT`) will use their standard column names
+(i.e. `text`).
+
+Alternatively, if the `INCLUDE KEY AS RECORD <name>` syntax is used all fields
+will be included as fields on a record with the specified name.
+
+If any key column names colide with value columns using the bare `INCLUDE KEY`
+syntax, an error will be raised before dataflow construction suggesting the use
+of the `AS RECORD` clause.
+
+### Implementation
+
+Mostly standard for new syntax additions. We will need to wait until
+`purify_format` to provide any of the specific error-detection because of the
+fact that Key schemas may come from the Confluent Schema Registry.
+
+Physically unpacking key fields will happen identically independent of if `AS
+RECORD` is specified, only the projection mapping and names available to SQL
+will differ.
+
+## Alternatives
```

Comment:
It seems like another alternative (unless I missed it) is "always include the key" and then rely on the *use* (or not) of that column to drive the instantiation of the key. Is it important for the source declaration to decide whether or not the key will be present in the source?

Reply:
That is a legitimate alternative! I'll include it. The largest downsides that I see to it are:

* `SELECT *`-type queries will change their semantics, which is a compatibility-impacting change (especially undesirable for things like debezium, where the key is afaik always a duplicated subset of the value fields)
* More generally, I think that for this to be an unambiguous API we would require that key and value were always presented as the `key`/`value` record syntax, which would be both more annoying and an even larger backcompat change.

Reply:
My sense is (maybe wrong) that internally we'll probably end up doing something like this, and the `INCLUDE KEY` stuff will desugar to the `map` statements that put the fields in the right place. At least, that seems to be the source orthogonality direction. And with that in mind, I have no strong opinion on how the user should be obliged to interact with it!

Reply:
I think that what you said is correct, if users spell out `INCLUDE KEY` but then don't actually use it we'll still elide it.